### PR TITLE
Refactor Auth modules to use UserEntityInterface/UserServiceInterface

### PR DESCRIPTION
**Pull request by demiankatz**
This PR refactors the authentication modules to use UserEntityInterface/UserServiceInterface instead of direct Laminas logic.

As of this writing, it is still work in progress, but it is currently in a stable state with a passing test suite.

TODO
- [x] Merge #3629 first and resolve conflicts
- [x] Merge #3723 and resolve conflicts
- [x] Eliminate all direct property accesses in favor of appropriate getters/setters
- [x] Refactor calls to VuFind\Db\Row\User methods that do not belong in UserEntityInterface
- [x] Add missing UserEntityInterface setters (to fix broken test)
- [x] Test all authentication methods (where possible)
- [x] Update changelog when merging

### CODE REVIEW

File: module/VuFind/src/VuFind/Auth/SIP2.php [hunk 1]
```diff
@@ -117,9 +118,6 @@ public function authenticate($request)
         ) {
             // Success!!!
             $user = $this->processSIP2User($result, $username, $password);
-
-            // Set login cookie for 1 hour
-            $user->password = $password; // Need this for Metalib
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Auth/SIP2.php:122]: I don't know what the point of this was in the distant past, but I'm confident it no longer has any effect and should be deleted. Whether this authentication method works at all is a separate question; I doubt it has been used in the past decade or more.

**Comment by demiankatz** [comment]
I believe this is ready for review, but I haven't been able to test every authentication driver. Here are some notes on the status of testing (and I'll edit them if and when I make progress on testing more scenarios):

  * AlmaDatabase - I cannot test this
  * CAS - this has unit tests, which are passing; I can't do hands-on testing
  * ChoiceAuth - this has unit and Mink test coverage, and all tests are passing
  * Database - this has unit and integration tests and is also used by the Mink test suite, and all tests are passing
  * Email - I cannot easily test this (though merging #3729 would help with that)
  * Facebook - this has been broken for a long time (see [VUFIND-1237](https://openlibraryfoundation.atlassian.net/browse/VUFIND-1237)); I'm maintaining the code in case anybody cares enough to eventually fix it
  * ILS - this has unit, integration and Mink tests, all of which pass
  * LDAP - this has unit tests, which are passing; I can't do hands-on testing
  * MultiAuth - this has unit tests, which are passing
  * MultiILS - this has unit tests, which are passing
  * PasswordAccess - this had no test coverage, so I've added a Mink test here to cover it -- it passes
  * Shibboleth - this has integration tests, all of which pass
  * SimulatedSSO - this has Mink test coverage, and all tests are passing
  * SIP2 - I'm skeptical that this still works (the code has been untouched for many years) but have no means of testing it; in any case, it has a passing unit test

File: module/VuFind/src/VuFind/Auth/AbstractBase.php [hunk 2]
```diff
@@ -533,4 +538,38 @@ protected function validateStringAgainstPolicy(
...
+     * @param string $username Username
+     *
+     * @return UserEntityInterface
+     * @throws Exception
+     */
+    protected function getOrCreateUserByUsername(string $username): UserEntityInterface
+    {
+        $userService = $this->getUserService();
+        $user = $userService->getUserByField('username', $username);
+        return $user ? $user : $userService->createEntity()->setUsername($username);
```
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Auth/AbstractBase.php]: Should this use the createEntityForUsername method?

Comment (demiankatz) [reply module/VuFind/src/VuFind/Auth/AbstractBase.php]: Looks like an oversight to me. I've changed it and tests are still passing. Thanks!

**Comment by EreMaijala** [comment]
I've tested ILS + Email authentication and didn't find any issues.

**Review by EreMaijala — APPROVED** [review]

**Pull request merged**